interfaces/devdoc/file_requirements.md

```diff
@@ -0,0 +1,143 @@
+MU_DEFINE_ENUM(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_VALUES);
+
+typedef struct FILE_HANDLE_DATA_TAG* FILE_HANDLE;
+typedef void(*FILE_REPORT_FAULT)(void* user_report_fault_context, const char* information);
+
+typedef void(*FILE_WRITE_CB)(void* user_context, bool is_successful);
+typedef void(*FILE_READ_CB)(void* user_context, bool is_successful, CONSTBUFFER_HANDLE content);
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create,FILE_REPORT_FAULT, user_report_fault_callback, void*, user_report_fault_context, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name);
+MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle);
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, CONSTBUFFER_HANDLE, source, uint64_t, position, FILE_WRITE_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_READ_ASYNC_RESULT, file_read_async, FILE_HANDLE, handle, uint32_t, size, uint64_t, position, FILE_READ_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+MOCKABLE_FUNCTION_WITH_RETURNS(, int, file_extend_filesize, FILE_HANDLE, handle, uint64_t, desired_size, bool, has_manage_volume)(0, MU_FAILURE);
+```
+
+## file_create
+
+```c
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create,FILE_REPORT_FAULT, user_report_fault_callback, void*, user_report_fault_context, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name);
+```
+
+`file_create` creates a file by the name of `full_file_name` it doesn't exist. `file_create` opens a file by the name `full_file_name` and returns its file handle.
+
+**SRS_FILE_43_033: [** If `execution_engine` is `NULL`, `file_create` shall fail and return `NULL`. **]**
+
+**SRS_FILE_43_002: [** If `full_file_name` is `NULL` then `file_create` shall fail and return `NULL`. **]**
+
+**SRS_FILE_43_037: [** If `full_file_name` is an empty string, `file_create` shall fail and return `NULL`. **]**
+
+**SRS_FILE_43_003: [** If a file with name `full_file_name` does not exist, `file_create` shall create a file with that name.**]**
+
+**SRS_FILE_43_001: [** `file_create` shall open the file named `full_file_name` for asynchronous operations and return its handle. **]**
+
+**SRS_FILE_43_034: [** If there are any failures, `file_create` shall fail and return `NULL`. **]**
+
+## file_destroy 
+
+```c
+MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle);
+```
+
+`file_destroy` closes the given file handle.
+
+**SRS_FILE_43_005: [** If `handle` is `NULL`, `file_destroy` shall return. **]**
+
+**SRS_FILE_43_006: [** `file_destroy` shall wait for all pending I/O operations to complete. **]**
+
+**SRS_FILE_43_007: [** `file_destroy` shall close the file handle `handle`. **]**
+
+## file_write_async
+
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, CONSTBUFFER_HANDLE, source, uint64_t, position, FILE_WRITE_CB, user_callback, void*, user_context)(0, MU_FAILURE);

```

Comment:
>position [](start = 135, length = 8)

if position + source->size exceed the file size, what's the behavior?

1) request denied
2) file is extended
3) write overflows to the beginning of the file?

Also, in case of "2" can the file end up extended but the data is not written?

Reply:
The file is extended. I am not sure whether the extension and write atomic or not. From the Linux specs, it does seem like they are. I couldn't find anything on the Windows spec.

---
In reply to: [450461074](https://github.com/Azure/azure-c-pal/pull/7#discussion_r450461074) [](ancestors = 450461074)